[Hub Generated] Review request for Microsoft.DBforPostgreSQL to add version preview/2023-03-02-preview

senatol opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://openapihub.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/senatol-postgresqlhsc-Microsoft.DBforPostgreSQL-2023-03-02-preview...main/).
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/b321e1dc-5e07-4940-9cd6-a65ac6729a8b)

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

senatol committed:
Adds base for updating Microsoft.DBforPostgreSQL from version stable/2022-11-08 to version 2023-03-02-preview

senatol committed:
Updates readme

senatol committed:
Updates API version in new specs and examples

senatol commented:
/azp run unifiedPipeline

senatol commented:
[MISSING_APIS_IN_DEFAULT_TAG](https://github.com/Azure/azure-rest-api-specs/pull/25509/checks?check_run_id=16176103655) errors in Swagger Avocado pipeline are due to namespace changing from "Microsoft.DBForPostgre**Sql**" in 2020-10-05-privatepreview version to "Microsoft.DBforPostgre**SQL**" which the latter one is the correct version which's been used in both stable 2022-11-08 version the version we'll publish currently, 2023-03-02-preview.

senatol committed:
Add database name property to cluster

senatol committed:
Add role properties

senatol committed:
Update role properties

senatol committed:
Add custom word entra

senatol committed:
Update enum

senatol committed:
Fix prettier

senatol committed:
Fix errors

senatol committed:
Revert change

senatol commented:
Hello @rkmanda ,

[MISSING_APIS_IN_DEFAULT_TAG](https://github.com/Azure/azure-rest-api-specs/pull/25509/checks?check_run_id=16176103655) errors in **Swagger Avocado** pipeline are due to namespace changing from "Microsoft.DBForPostgre**Sql**" in 2020-10-05-privatepreview version to "Microsoft.DBforPostgre**SQL**". 
The 2020-10-05-privatepreview version was mistakenly pushed to this repository and was not released to customers. The latter one is the correct version which's been used in both stable 2022-11-08 version the version we'll publish currently, 2023-03-02-preview.

TimLovellSmith commented on custom-words.txt:
```diff
@@ -847,6 +847,7 @@ enquoted
 enrollmentgroups
 entitydef
 entitysearch
+entra
```
Can we remove it?

TimLovellSmith commented:
>         "entraIdPrincipal": {

I think we shouldn't put the brand 'entra' in the API. Rebranding will make the API naming age ungracefully...

Instead, let's consider we can call it something more like 'identity'

---
Refers to: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/preview/2023-03-02-preview/postgresqlhsc.json:2621 in 4caf250. [](commit_id = 4caf2507ca2d5cb006c09f4d2c6695dc47687c78, deletion_comment = False)

TimLovellSmith commented:
>         },

Please note that booleans are hard to extend and not future proof, so discouraged. Consider using an enum or other optional property instead.

---
Refers to: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/preview/2023-03-02-preview/postgresqlhsc.json:2610 in 4caf250. [](commit_id = 4caf2507ca2d5cb006c09f4d2c6695dc47687c78, deletion_comment = False)

TimLovellSmith commented:
>         "password": {

Great to see it won't be required!
How will someone know whether it was ever set or not though? Since you can't see a password set in the response.
Consider having a flag to say whether password auth is enabled or disabled, or just documenting the fact if its always implicitly disabled any time an identity is used.

---
Refers to: specification/postgresqlhsc/resource-manager/Microsoft.DBforPostgreSQL/preview/2023-03-02-preview/postgresqlhsc.json:2611 in 4caf250. [](commit_id = 4caf2507ca2d5cb006c09f4d2c6695dc47687c78, deletion_comment = False)

senatol committed:
Add group type support

senatol committed:
Change entraIdPrincipal to externalIdentity

senatol committed:
Change boolean property to enum

senatol committed:
Update format

senatol commented:
Hello @TimLovellSmith,
I applied recommended changes to the PR. I also updated the resource manager backend code in the meantime.
- Changed `isAdmin` boolean type to enum `roleType`
- Updated description for password to emphasize if an identity is used, password will not be required&used.
- Changed `entraIdPrincipal `to `externalIdentity` to make if future proof.

Thank you for your review!

senatol committed:
Change uuid format to arm-id

senatol committed:
Add promote body for enableGeoBackup
